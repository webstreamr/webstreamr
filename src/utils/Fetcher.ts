import CachePolicy from 'http-cache-semantics';
import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, TIMEOUT } from '../types';
import { BlockedError, NotFoundError, QueueIsFullError } from '../error';
import { clearTimeout } from 'node:timers';

interface HttpCacheItem {
  policy: CachePolicy;
  status: number;
  statusText: string;
  body: string;
}

interface HostQueue {
  promise: Promise<void>;
  count: number;
}

export class Fetcher {
  private readonly logger: winston.Logger;
  private readonly httpCache: TTLCache<string, HttpCacheItem>;
  private readonly hostQueues = new Map<string, HostQueue>();

  constructor(logger: winston.Logger) {
    this.logger = logger;
    this.httpCache = new TTLCache();
  }

  readonly text = async (ctx: Context, url: URL, init?: RequestInit): Promise<string> => {
    return (await this.cachedFetch(ctx, url, init)).body;
  };

  readonly textPost = async (ctx: Context, url: URL, data: unknown, init?: RequestInit): Promise<string> => {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'POST', body: JSON.stringify(data) })).body;
  };

  readonly head = async (ctx: Context, url: URL, init?: RequestInit): Promise<CachePolicy.Headers> => {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'HEAD' })).policy.responseHeaders();
  };

  public readonly getInit = (ctx: Context, url: URL, init?: RequestInit): RequestInit => ({
    ...init,
    headers: {
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'en',
      'Forwarded': `for=${ctx.ip}`,
      'Priority': 'u=0',
      'Referer': `${ctx.referer?.href ?? url.origin}`,
      'User-Agent': 'node',
      'X-Forwarded-For': ctx.ip,
      'X-Forwarded-Proto': url.protocol.slice(0, -1),
      'X-Real-IP': ctx.ip,
      ...init?.headers,
    },
  });

  private readonly handleHttpCacheItem = (httpCacheItem: HttpCacheItem): HttpCacheItem => {
    if (httpCacheItem.status && httpCacheItem.status >= 200 && httpCacheItem.status <= 299) {
      return httpCacheItem;
    }

    if (httpCacheItem.status === 404) {
      throw new NotFoundError();
    }

    const responseHeaders = httpCacheItem.policy.responseHeaders();

    if (httpCacheItem.policy.responseHeaders()['cf-mitigated'] === 'challenge') {
      throw new BlockedError('cloudflare_challenge');
    }

    if (httpCacheItem.status === 403) {
      throw new BlockedError('unknown');
    }

    throw new Error(`Fetcher error: ${httpCacheItem.status}: ${httpCacheItem.statusText}, response headers: ${JSON.stringify(responseHeaders)}`);
  };

  private readonly headersToObject = (headers: Headers): Record<string, string> => {
    const obj: Record<string, string> = {};

    headers.forEach((value, name) => {
      obj[name] = value;
    });

    return obj;
  };

  private readonly cachedFetch = async (ctx: Context, url: URL, init?: RequestInit): Promise<HttpCacheItem> => {
    const newInit = this.getInit(ctx, url, init);

    const request: CachePolicy.Request = { url: url.href, method: newInit.method ?? 'GET', headers: {} };

    let httpCacheItem: HttpCacheItem | undefined = this.httpCache.get(url.href);
    if (httpCacheItem) {
      this.logger.info(`Cached fetch ${request.method} ${url}`, ctx);
      return this.handleHttpCacheItem(httpCacheItem);
    }

    const response = await this.queuedFetch(ctx, url, newInit);
    const body = await response.text();

    const policy = new CachePolicy(request, { status: response.status, headers: this.headersToObject(response.headers) }, { shared: false });

    httpCacheItem = { policy, status: response.status, statusText: response.statusText, body };

    let ttl;
    if (response.ok) {
      ttl = Math.max(policy.timeToLive(), 900000); // 15m at least
    } else {
      ttl = policy.timeToLive();
    }

    if (ttl > 0) {
      this.httpCache.set(url.href, httpCacheItem, { ttl });
    }

    return this.handleHttpCacheItem(httpCacheItem);
  };

  private readonly queuedFetch = async (ctx: Context, url: URL, init?: RequestInit): Promise<Response> => {
    let queue = this.hostQueues.get(url.host);
    if (!queue) {
      queue = { promise: Promise.resolve(), count: 0 };
      this.hostQueues.set(url.host, queue);
    }

    queue.count++;

    if (queue.count > 10) {
      throw new QueueIsFullError();
    }

    let resolveCurrent: () => void;
    const currentPromise = new Promise<void>(
      (resolve) => { resolveCurrent = resolve; },
    );

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(TIMEOUT), 10000);

    const fetchPromise = queue.promise.then(async () => {
      try {
        this.logger.info(`Fetch ${init?.method ?? 'GET'} ${url}`, ctx);

        return await fetch(url, { ...init, signal: controller.signal });
      } finally {
        clearTimeout(timer);
        resolveCurrent();
        queue.count--;
      }
    });

    queue.promise = currentPromise.catch(/* istanbul ignore next */ () => undefined);

    return fetchPromise;
  };
}
