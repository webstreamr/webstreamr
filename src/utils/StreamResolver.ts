import { Stream } from 'stremio-addon-sdk';
import { flag } from 'country-emoji';
import winston from 'winston';
import bytes from 'bytes';
import { Context, TIMEOUT, UrlResult } from '../types';
import { Handler } from '../handler';
import { BlockedError, NotFoundError, QueueIsFullError } from '../error';
import { languageFromCountryCode } from './languageFromCountryCode';

export class StreamResolver {
  private readonly logger: winston.Logger;

  constructor(logger: winston.Logger) {
    this.logger = logger;
  }

  readonly resolve = async (ctx: Context, handlers: Handler[], type: string, id: string): Promise<Stream[]> => {
    if (handlers.length === 0) {
      return [{
        name: 'WebStreamr',
        title: '⚠️ No handlers found. Please re-configure the plugin.',
        ytId: 'E4WlUXrJgy4',
      }];
    }

    const streams: Stream[] = [];

    const urlResults: UrlResult[] = [];
    const handlerPromises = handlers.map(async (handler) => {
      if (!handler.contentTypes.includes(type)) {
        return;
      }

      try {
        const handlerUrlResults = await handler.handle(ctx, type, id);
        this.logger.info(`${handler.id} returned ${handlerUrlResults.length} urls`, ctx);

        urlResults.push(...(handlerUrlResults.filter(handlerUrlResult => handlerUrlResult !== undefined)));
      } catch (error) {
        if (error instanceof NotFoundError) {
          return;
        }

        streams.push({
          name: process.env['MANIFEST_NAME'] || 'WebStreamr',
          title: [`🔗 ${handler.label}`, this.logErrorAndReturnNiceString(ctx, handler.id, error)].join('\n'),
          ytId: 'E4WlUXrJgy4',
        });
      }
    });
    await Promise.all(handlerPromises);

    urlResults.sort((a, b) => {
      const heightComparison = (b.meta.height ?? 0) - (a.meta.height ?? 0);
      if (heightComparison !== 0) {
        return heightComparison;
      }

      const bytesComparison = (b.meta.bytes ?? 0) - (a.meta.bytes ?? 0);
      if (bytesComparison !== 0) {
        return bytesComparison;
      }

      return a.label.localeCompare(b.label);
    });

    this.logger.info(`Return ${urlResults.length} streams`, ctx);

    streams.push(
      ...urlResults.filter(urlResult => !urlResult.isExternal || this.showExternalUrls(ctx) || urlResult.error)
        .map(urlResult => ({
          ...this.buildUrl(ctx, urlResult),
          name: this.buildName(ctx, urlResult),
          title: this.buildTitle(ctx, urlResult),
          behaviorHints: {
            ...(urlResult.sourceId && { bingeGroup: `webstreamr-${urlResult.sourceId}` }),
            ...(urlResult.requestHeaders !== undefined && {
              notWebReady: true,
              proxyHeaders: { request: urlResult.requestHeaders },
            }),
            ...(urlResult.meta.bytes && { videoSize: urlResult.meta.bytes }),
          },
        })),
    );

    return streams;
  };

  private readonly showExternalUrls = (ctx: Context): boolean => !('excludeExternalUrls' in ctx.config);

  private readonly buildUrl = (ctx: Context, urlResult: UrlResult): { externalUrl: string } | { url: string } | { ytId: string } => {
    if (!urlResult.isExternal) {
      return { url: urlResult.url.href };
    }

    if (this.showExternalUrls(ctx)) {
      return { externalUrl: urlResult.url.href };
    }

    return { ytId: 'E4WlUXrJgy4' };
  };

  private readonly buildName = (ctx: Context, urlResult: UrlResult): string => {
    let name = process.env['MANIFEST_NAME'] || 'WebStreamr';

    name += urlResult.meta.height ? ` ${urlResult.meta.height}P` : ' N/A';

    if (urlResult.isExternal && this.showExternalUrls(ctx)) {
      name += ` ⚠️ external`;
    }

    return name;
  };

  private readonly logErrorAndReturnNiceString = (ctx: Context, source: string, error: unknown): string => {
    if (error instanceof BlockedError) {
      this.logger.warn(`${source}: Request was blocked. Reason: ${error.reason}`, ctx);

      return '⚠️ Request was blocked.';
    }

    if (error === TIMEOUT) {
      this.logger.warn(`${source}: Request timed out.`, ctx);

      return '🐢 Request timed out.';
    }

    if (error instanceof QueueIsFullError) {
      this.logger.warn(`${source}: Request queue is full.`, ctx);

      return '⏳ Request queue is full. Please try again later or consider self-hosting.';
    }

    const cause = (error as Error & { cause?: unknown }).cause;
    this.logger.error(`${source} error: ${error}, cause: ${cause}`, ctx);

    return `❌ Request failed. Request-id: ${ctx.id}.`;
  };

  private readonly buildTitle = (ctx: Context, urlResult: UrlResult): string => {
    const titleLines = [];

    if (urlResult.meta.title) {
      titleLines.push(`📂 ${urlResult.meta.title}`);
    }

    if (urlResult.meta.bytes) {
      titleLines.push(`💾 ${bytes.format(urlResult.meta.bytes, { unitSeparator: ' ' })}`);
    }

    if (urlResult.meta.countryCode) {
      titleLines.push(`🌐 ${languageFromCountryCode(urlResult.meta.countryCode)} ${flag(urlResult.meta.countryCode)}`);
    }

    titleLines.push(`🔗 ${urlResult.label}`);

    if (urlResult.error) {
      titleLines.push(this.logErrorAndReturnNiceString(ctx, urlResult.sourceId, urlResult.error));
    }

    return titleLines.join('\n');
  };
}
