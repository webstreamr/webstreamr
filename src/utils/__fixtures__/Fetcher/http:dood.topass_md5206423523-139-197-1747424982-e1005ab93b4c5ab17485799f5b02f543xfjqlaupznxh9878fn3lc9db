https://sc582x.cloudatacdn.com/u5kj7hvlixblsdgge5a4eza2jq47t2hxw6eluvwfaatnwq6j4skdwj4oglea/o59vueq2rh~