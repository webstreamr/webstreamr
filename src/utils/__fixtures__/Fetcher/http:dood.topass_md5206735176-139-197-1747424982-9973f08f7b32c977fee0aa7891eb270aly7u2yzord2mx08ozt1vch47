https://dw120l.cloudatacdn.com/u5kj6npjllblsdgge6dmeyqoldb72gsikji45jrtdcluvvchjphpotxq7uyq/xqe2dvlb2w~