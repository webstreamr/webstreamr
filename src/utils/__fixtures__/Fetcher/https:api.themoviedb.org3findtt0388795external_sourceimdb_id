{"movie_results":[{"backdrop_path":"/vXzqV2RtJlT9WhNEM4NYWCEC6ZI.jpg","id":142,"title":"Brokeback Mountain","original_title":"Brokeback Mountain","overview":"In 1960s Wyoming, two men develop a strong emotional and sexual relationship that endures as a lifelong connection complicating their lives as they get married and start families of their own.","poster_path":"/aByfQOQBNa4CMFwIgq3QrqY2ZHh.jpg","media_type":"movie","adult":false,"original_language":"en","genre_ids":[18,10749],"popularity":8.5469,"release_date":"2005-10-22","video":false,"vote_average":7.8,"vote_count":7169}],"person_results":[],"tv_results":[],"tv_episode_results":[],"tv_season_results":[]}