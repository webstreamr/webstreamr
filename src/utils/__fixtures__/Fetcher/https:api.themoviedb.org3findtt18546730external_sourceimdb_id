{"movie_results":[],"person_results":[],"tv_results":[{"backdrop_path":"/vV5LKWmuysEe5wsuZJGbdiL5XJ2.jpg","id":194583,"name":"The Walking Dead: Dead City","original_name":"The Walking Dead: Dead City","overview":"Maggie and Negan travel to post-apocalyptic Manhattan - long ago cut off from the mainland. The crumbling city is filled with the dead and denizens who have made it a world full of anarchy, danger, beauty, and terror.","poster_path":"/wq3vuQzQgbS83zX3malAFWMsSwX.jpg","media_type":"tv","adult":false,"original_language":"en","genre_ids":[10759,18,10765],"popularity":54.1718,"first_air_date":"2023-06-18","vote_average":8.017,"vote_count":609,"origin_country":["US"]}],"tv_episode_results":[],"tv_season_results":[]}