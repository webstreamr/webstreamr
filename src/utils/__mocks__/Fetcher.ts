import fs from 'node:fs';
import Axios, { AxiosRequestConfig } from 'axios';
import slugify from 'slugify';
import { Context } from '../../types';

export class Fetcher {
  readonly text = async (_ctx: Context, url: URL, config?: AxiosRequestConfig): Promise<string> => {
    const path = `${__dirname}/../__fixtures__/Fetcher/${slugify(url.href)}`;

    if (fs.existsSync(path)) {
      return fs.readFileSync(path).toString();
    } else {
      const text = (await Axios.create().get(url.href, config)).data;

      fs.writeFileSync(path, text);

      return text;
    }
  };

  readonly textPost = async (_ctx: Context, url: URL, data: unknown, config?: AxiosRequestConfig): Promise<string> => {
    const path = `${__dirname}/../__fixtures__/Fetcher/post-${slugify(url.href)}-${slugify(JSON.stringify(data))}`;

    if (fs.existsSync(path)) {
      return fs.readFileSync(path).toString();
    } else {
      const text = (await Axios.create().post(url.href, data, config)).data;

      fs.writeFileSync(path, text);

      return text;
    }
  };

  public readonly getUserAgentForIp = (): string => {
    return 'Fake UserAgent';
  };
}
