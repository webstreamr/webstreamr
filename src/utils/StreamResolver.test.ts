import winston from 'winston';
import { ExtractorRegistry } from '../extractor';
import { StreamResolver } from './StreamResolver';
import { Handler, MeineCloud, MostraGuarda } from '../handler';
import { Fetcher } from './Fetcher';
import { Context, CountryCode, TIMEOUT, UrlResult } from '../types';
import { BlockedError, NotFoundError, QueueIsFullError } from '../error';
jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const streamResolver = new StreamResolver(logger);
const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { de: 'on', it: 'on' } };

// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const extractorRegistry = new ExtractorRegistry(logger, fetcher);
const meineCloud = new MeineCloud(fetcher, extractorRegistry);
const mostraGuarda = new MostraGuarda(fetcher, extractorRegistry);

describe('resolve', () => {
  test('returns info as stream if no handlers were configured', async () => {
    const streams = await streamResolver.resolve(ctx, [], 'movie', 'tt123456789');

    expect(streams).toStrictEqual([{
      name: 'WebStreamr',
      title: '⚠️ No handlers found. Please re-configure the plugin.',
      ytId: 'E4WlUXrJgy4',
    }]);
  });

  test('returns handler errors as stream', async () => {
    const fetcherSpy = jest.spyOn(fetcher, 'text').mockRejectedValue('ups, an error occurred.');

    const streams = await streamResolver.resolve(ctx, [meineCloud], 'movie', 'tt123456789');

    expect(streams).toStrictEqual([{
      name: 'WebStreamr',
      title: '🔗 MeineCloud\n❌ Request failed. Request-id: id.',
      ytId: 'E4WlUXrJgy4',
    }]);

    fetcherSpy.mockRestore();
  });

  test('returns empty array if no handler found anything', async () => {
    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'movie', 'tt12345678');
    expect(streams).toStrictEqual([]);
  });

  test('returns empty array if no handler supported the type', async () => {
    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'series', 'tt12345678:1:1');
    expect(streams).toStrictEqual([]);
  });

  test('returns sorted results', async () => {
    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'movie', 'tt29141112');
    expect(streams).toMatchSnapshot();
  });

  test('adds error info', async () => {
    class MockHandler implements Handler {
      readonly id = 'mockhandler';

      readonly label = 'MockHandler';

      readonly contentTypes = ['movie'];

      readonly countryCodes: CountryCode[] = ['de'];

      readonly handle = async (): Promise<(UrlResult | undefined)[]> => {
        return [
          {
            url: new URL('https://example.com'),
            isExternal: true,
            error: new BlockedError('cloudflare_challenge'),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: 'de',
            },
          },
          {
            url: new URL('https://example2.com'),
            isExternal: true,
            error: new TypeError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: 'de',
            },
          },
          {
            url: new URL('https://example2.com'),
            isExternal: true,
            error: TIMEOUT,
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: 'de',
            },
          },
          {
            url: new URL('https://example3.com'),
            isExternal: true,
            error: new QueueIsFullError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: 'de',
            },
          },
        ];
      };
    }

    const streams = await streamResolver.resolve(ctx, [new MockHandler()], 'movie', 'tt11655566');
    expect(streams).toMatchSnapshot();

    const streamsWithoutExternalUrls = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, excludeExternalUrls: 'on' } }, [new MockHandler()], 'movie', 'tt11655566');
    expect(streamsWithoutExternalUrls).toMatchSnapshot();
  });

  test('ignores not found errors', async () => {
    const mockHandler: Handler = {
      id: 'mockhandler',
      label: 'MockHandler',
      contentTypes: ['movie'],
      countryCodes: ['de'],
      handle: jest.fn().mockRejectedValue(new NotFoundError()),
    };

    const streams = await streamResolver.resolve(ctx, [mockHandler], 'movie', 'tt12345678');
    expect(streams).toStrictEqual([]);
  });
});
