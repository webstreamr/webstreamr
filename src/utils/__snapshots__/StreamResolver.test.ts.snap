// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve adds error info 1`] = `
[
  {
    "behaviorHints": {},
    "externalUrl": "https://example.com/",
    "name": "WebStreamr N/A ⚠️ external",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
⚠️ Request was blocked.",
  },
  {
    "behaviorHints": {},
    "externalUrl": "https://example2.com/",
    "name": "WebStreamr N/A ⚠️ external",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
❌ Request failed. Request-id: id.",
  },
  {
    "behaviorHints": {},
    "externalUrl": "https://example2.com/",
    "name": "WebStreamr N/A ⚠️ external",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
🐢 Request timed out.",
  },
  {
    "behaviorHints": {},
    "externalUrl": "https://example3.com/",
    "name": "WebStreamr N/A ⚠️ external",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
⏳ Request queue is full. Please try again later or consider self-hosting.",
  },
]
`;

exports[`resolve adds error info 2`] = `
[
  {
    "behaviorHints": {},
    "name": "WebStreamr N/A",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
⚠️ Request was blocked.",
    "ytId": "E4WlUXrJgy4",
  },
  {
    "behaviorHints": {},
    "name": "WebStreamr N/A",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
❌ Request failed. Request-id: id.",
    "ytId": "E4WlUXrJgy4",
  },
  {
    "behaviorHints": {},
    "name": "WebStreamr N/A",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
🐢 Request timed out.",
    "ytId": "E4WlUXrJgy4",
  },
  {
    "behaviorHints": {},
    "name": "WebStreamr N/A",
    "title": "🌐 German 🇩🇪
🔗 hoster.com
⏳ Request queue is full. Please try again later or consider self-hosting.",
    "ytId": "E4WlUXrJgy4",
  },
]
`;

exports[`resolve returns sorted results 1`] = `
[
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_de",
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 1080P",
    "title": "📂 des-teufels-bad-2024
💾 1.3 GB
🌐 German 🇩🇪
🔗 Dropload",
    "url": "https://srv27.dropload.io/hls2/01/00197/lyo2h1snpe5c_h/master.m3u8?t=qTZsRGIMnUjXsQ4o2LRLlsT4tGG1AdJrRBRDGttNLPM&s=1746903744&e=14400&f=987607&i=0.0&sp=0&ii=130.61.236.204",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_it",
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 720P",
    "title": "📂 the-devils-bath-2024-sd-sub-ita
💾 1.1 GB
🌐 Italian 🇮🇹
🔗 Dropload",
    "url": "https://srv34.dropload.io/hls2/01/00200/xsr90y2ltdyx_h/master.m3u8?t=avFOZ-10hDxo6__iaoSUrq7o4yY3ldLw6XDlXxpXF8E&s=1747059001&e=14400&f=1004116&srv=srv24&i=0.0&sp=0&ii=130.61.236.204&p1=srv24&p2=srv24",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_it",
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 720P",
    "title": "📂 the devils bath 2024 sub ita
💾 1.1 GB
🌐 Italian 🇮🇹
🔗 SuperVideo",
    "url": "https://hfs279.serversicuro.cc/hls/,dnzpejj427g4a3gyvbth53rxqewqgqqzp3wr6zwutysvud7zroar3cb75d3q,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_de",
      "videoSize": 1073741824,
    },
    "name": "WebStreamr 720P",
    "title": "📂 des teufels bad 2024
💾 1 GB
🌐 German 🇩🇪
🔗 SuperVideo",
    "url": "https://hfs290.serversicuro.cc/hls/,dnzpfi3d27g4a3gyvbmh5klwtl65qh654hyjtgupd6p56thhyqu6ra3olbfa,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_de",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to/",
        },
      },
    },
    "name": "WebStreamr N/A",
    "title": "📂 des-teufels-bad-2024
🌐 German 🇩🇪
🔗 DoodStream",
    "url": "https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rtpmbzgb5tornooa/k9wk0js17e~mocked-random-string?token=7uebebipnnhusa4xnyea1er4&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_it",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to/",
        },
      },
    },
    "name": "WebStreamr N/A",
    "title": "📂 the-devils-bath-2024-sd-sub-ita
🌐 Italian 🇮🇹
🔗 DoodStream",
    "url": "https://kk892as.cloudatacdn.com/u5kj7j7s27d3sdgge5woezkbi4pu672wktq3aqujw47rbpl3xog2gtqnw2xa/awn9rgotuu~mocked-random-string?token=aw2v2d8uhbrj2ky54d573ujk&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-external_de",
    },
    "externalUrl": "https://mixdrop.ag/e/3nzwveprim63or6",
    "name": "WebStreamr N/A ⚠️ external",
    "title": "🌐 German 🇩🇪
🔗 mixdrop.ag",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-external_it",
    },
    "externalUrl": "https://mixdrop.ag/e/vk196d6xfzwwo1",
    "name": "WebStreamr N/A ⚠️ external",
    "title": "🌐 Italian 🇮🇹
🔗 mixdrop.ag",
  },
]
`;
